### PR TITLE
rootless: expose ports automatically

### PR DESCRIPTION
**Pull request by AkihiroSuda**
Signed-off-by: Akihiro Suda <suda.akihiro@lab.ntt.co.jp>

**- What I did**

Now `docker run -p` ports can be exposed to the host namespace automatically when `dockerd-rootless.sh` is launched with `--userland-proxy --userland-proxy-path $(which rootlesskit-docker-proxy)`.
This is akin to how Docker for Mac/Win works with `--userland-proxy-path=/path/to/vpnkit-expose-port`.

The port number on the host namespace needs to be set to >= 1024.
SCTP ports are currently unsupported.

New binaries built from RootlessKit repo:
* `rootlesskit-docker-proxy`: drop-in replacement for `docker-proxy`.  Needed for exposing ports.
* ~`rootlessctl`: RootlessKit REST API client. Added only for ease of debugging, as in `ctr`.~

RootlessKit changes: https://github.com/rootless-containers/rootlesskit/compare/7bbbc48a6f906633a9b12783b957f4c3aa037d33...ed2671442965115b84ecf82d4831cc48747d89b8

**- How I did it**

By using `rootlesskit-docker-proxy` with  RootlessKit's `builtin` port driver.

* `rootlesskit-docker-proxy`: https://github.com/rootless-containers/rootlesskit/tree/v0.3.0-beta.0/cmd/rootlesskit-docker-proxy
* `builtin` port driver: https://github.com/rootless-containers/rootlesskit/blob/v0.3.0-beta.0/pkg/port/builtin/builtin.go

**- How to verify it**

```console
$ dockerd-rootless.sh --experimental --userland-proxy --userland-proxy-path=$(which rootlesskit-docker-proxy)
```

```console
$ export DOCKER_HOST=unix://$XDG_RUNTIME_DIR/docker.sock
$ docker run -d -p 8080:80 nginx:alpine
$ w3m http://localhost:8080
```

**- Description for the changelog**

rootless: expose ports automatically

**- A picture of a cute animal (not mandatory but encouraged)**

penguin

https://upload.wikimedia.org/wikipedia/commons/f/f5/Little_Penguin_Feb09.jpg

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
cc @tonistiigi @tiborvass @djs55

File: hack/make/binary-daemon [hunk 1]
```diff
@@ -14,7 +14,7 @@ copy_binaries() {
 		return
 	fi
 	echo "Copying nested executables into $dir"
-	for file in containerd containerd-shim ctr runc docker-init docker-proxy rootlesskit dockerd-rootless.sh; do
+	for file in containerd containerd-shim ctr runc docker-init docker-proxy rootlesskit rootlessctl rootlesskit-docker-proxy dockerd-rootless.sh; do
```
Comment (tonistiigi) [inline hack/make/binary-daemon]: q: could we avoid another binary for `rootlessctl` (for size) . Eg. could it be just a subcommand of rootlesskit?

Comment (AkihiroSuda) [reply hack/make/binary-daemon]: removed rootlessctl

**Comment by AkihiroSuda** [comment]
@tonistiigi can we get this in 19.03?

**Comment by AkihiroSuda** [comment]
@tonistiigi @tiborvass @thaJeztah @djs55 PTAL?

**Comment by thaJeztah** [comment]
I asked @tonistiigi and @tiborvass to have a look 😥🤗

**Comment by djs55** [comment]
Looks good to me!

**Review by thaJeztah — APPROVED** [review]
thanks @djs55. I'll copy your LGTM

but would love a quick look from @tonistiigi and/or @tiborvass :smile:

File: docs/rootless.md [hunk 2]
```diff
@@ -45,6 +45,8 @@ penguin:231072:65536
   * Apparmor
   * Checkpoint
   * Overlay network
+  * Exposing SCTP ports
+* To expose a TCP/UDP port, the host port number needs to be set to >= 1024.
```
Comment (thaJeztah) [inline docs/rootless.md]: oh; nit: looks like indentation is off here

Comment (AkihiroSuda) [reply docs/rootless.md]: This is intentional. `Exposing SCTP ports` is an unsupported feature while `To expose a TCP/UDP port...` is a restriction

Comment (thaJeztah) [reply docs/rootless.md]: oh! should've looked at the whole file; apologies

**Comment by tiborvass** [comment]
I couldn't get this to work inside the dev container.
```
$ docker run --name test -dit -p 1000:5000/tcp tiborvass/netcat nc -l -p 5000
9848d1fe9511328ebf84199fafd28f78e7bafdab1a6ddebce955398292e672bf
docker: Error response from daemon: driver failed programming external connectivity on endpoint test (9ff63a025c70566db7330602fced8c1759aba2f82c00cbf04b07a7c90b0562ff): Error starting userland proxy:.
```

Daemon logs:
```
time="2019-04-02T22:07:20.041164565Z" level=warning msg="Failed to allocate and map port 1000-1000: Error starting userland proxy: "
time="2019-04-02T22:07:20.104852910Z" level=warning msg="9848d1fe9511328ebf84199fafd28f78e7bafdab1a6ddebce955398292e672bf cleanup: failed to unmount IPC: umount /home/foo/.local/share/docker/containers/9848d1fe9511328ebf84199fafd28f78e7bafdab1a6ddebce955398292e672bf/mounts/shm, flags: 0x2: no such file or directory"
time="2019-04-02T22:07:20.105603920Z" level=error msg="9848d1fe9511328ebf84199fafd28f78e7bafdab1a6ddebce955398292e672bf cleanup: failed to delete container from containerd: no such container"
time="2019-04-02T22:07:20.105668562Z" level=error msg="Handler for POST /v1.30/containers/9848d1fe9511328ebf84199fafd28f78e7bafdab1a6ddebce955398292e672bf/start returned error: driver failed programming external connectivity on endpoint test (9ff63a025c70566db7330602fced8c1759aba2f82c00cbf04b07a7c90b0562ff): Error starting userland proxy: "
```

When I start rootlesskit-docker-proxy manually to test it, it says:
```
2019/04/02 22:09:23 $ROOTLESSKIT_STATE_DIR needs to be set
```

So I don't know if that's something one of your binaries are setting.

**Comment by AkihiroSuda** [comment]
> `-p 1000:5000/tcp tiborvass/netcat nc -l -p 5000`

You can't listen on a port number below 1024 unless you have CAP_NET_BIND_SERVICE

> `2019/04/02 22:09:23 $ROOTLESSKIT_STATE_DIR needs to be set`
Please make sure to have rootlesskit  v0.3.0-beta.0 or later?  https://github.com/rootless-containers/rootlesskit/commit/732a90a7cfe2bbed2f81832afe07c38940d45a14

**Comment by tiborvass** [comment]
@AkihiroSuda ok that worked. I tried with 2000.

I was testing inside the dev container from moby's Dockerfile, which should have the correct version of rootlesskit. It's just that I was running it manually from bash.

**Review by tiborvass — APPROVED** [review]

**Comment by AkihiroSuda** [comment]
19.03: https://github.com/docker/engine/pull/189

**Pull request merged**